synapse/util/caches/response_cache.py

```diff
@@ -187,6 +187,16 @@ def unset(self, key: KV) -> None:
         """
         self._result_cache.pop(key, None)
 
+    def evict_because(self, key: KV, reason: EvictionReason) -> None:
+        """Basically the same as unset, but update reason why evicting for metrics"""
+        self._metrics.inc_evictions(reason)
+        self.unset(key)
+
+    def entry_timeout(self, key: KV) -> None:
```

Comment:
This seems reasonable, but I have a few suggestions:

* Let's make this private (add a `_` in front of it).
* Get rid of the `evict_because` method -- it is trivial and doesn't add much IMO / only adds indirection.
* Stop calling `unset` and inline it.
* Update `unset` to include `self._metrics.inc_evictions(EvicationReason.invalidation)`.

Reply:
I think I got it all into [59563dd](https://github.com/matrix-org/synapse/pull/16028/commits/59563ddddad7f1bb4c59cf321e4334e496f20236)